src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -107,6 +107,11 @@ internal abstract partial class AbstractFlowPass<TLocalState, TLocalFunctionStat
         protected TLocalState StateWhenFalse;
         protected bool IsConditionalState;
 
+        /// <summary>
+        /// The state after a conditional access when the conditional access's value is non-null.
+        /// </summary>
+        protected TLocalState StateWhenNotNull;
```

Comment:
Note that this was originally implemented by adding an overload of 'VisitConversion' which had a parameter `out TLocalState stateWhenNotNull`, then calling it directly from methods like `VisitNullCoalescingOperator` when appropriate.

However, this approach ends up missing the code path in `VisitAlways` where we handle region analysis, so it broke an "extract method" test. It ended up feeling necessary for practical reasons to make it so "additional" outputs of visit methods are given by writing to fields instead of by overloading with additional outputs.

There are some concerns about the scope/lifetime of this field. It feels like ideally we could do something to make it so this is only populated when the caller wants it, and it is discarded shortly after the caller is done using it.

Alternatively we could try to prototype a scheme where we can "split" and "unsplit" to and from a null-conditional state, and then attempt to handle more scenarios where expressions are evaluated null-conditionally. For example, lifted operators: `a?.b(out x) && c?.d(out y)`. It feels like we are uncertain of the potential cost of this approach.

Reply:
>However, this approach ends up missing the code path in VisitAlways where we handle region analysis, so it broke an "extract method" test

I'm honestly concerned by this approach. Could we just do the region analysis bits we need to in `VisitNullCoalescingOperator`?

---
In reply to: [590761815](https://github.com/dotnet/roslyn/pull/51567#discussion_r590761815) [](ancestors = 590761815)

Reply:
Have made the change, thanks.